### PR TITLE
Feature/f 35+36+38

### PR DESCRIPTION
**Pull request by FiryDragon1**
Added SMSListener.
The app now responds to SMS it recieves.
Added the ability to query the current location of the device.

### CODE REVIEW

File: app/src/main/java/at/tacticaldevc/oat/utils/Prefs.java [hunk 1]
```diff
@@ -27,16 +34,125 @@
...
+        ensureNotNull(context, "Application Context");
+        try {
+            ensureStringIsValid(passwordToCheck, "password to check");
+        } catch (IllegalArgumentException ex) {
+            return false;
+        }
+
+        SharedPreferences prefs = context.getSharedPreferences(DOCUMENT_NAME_DATA, Context.MODE_PRIVATE);
+
+        String salt = prefs.getString(KEY_COMMAND_PASSWORD_SALT, "");
```
Comment (tacticalDevC) [inline app/src/main/java/at/tacticaldevc/oat/utils/Prefs.java]: IMO we should throw an exception if there is no salt saved. It doesn't make sense if there is a password but no salt. Having no salt saved would mean that the password is lost and the data must be exported/deleted

File: app/src/main/java/at/tacticaldevc/oat/utils/SMSCom.java [hunk 2]
```diff
@@ -1,7 +1,88 @@
...
+    /**
+     * @return the SMSManager associated with the default Subscription or null, if no valid subscription could be found
+     */
+    private static SmsManager fetchSMSManager() {
+        int subscriptionID = SmsManager.getDefaultSmsSubscriptionId();
+        if (!(subscriptionID == SubscriptionManager.INVALID_SUBSCRIPTION_ID)) {
+            return SmsManager.getSmsManagerForSubscriptionId(subscriptionID);
+        }
+        // Error handling missing
+        return null;
```
Comment (tacticalDevC) [inline app/src/main/java/at/tacticaldevc/oat/utils/SMSCom.java]: OATException.forLibraryError()

File: app/src/main/res/values/oat_sms_texts.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="oat_sms_message_error_feature_disabled">OAT: \n The feature %s is disabled and cannot be accessed at the moment. Enable this feature in the settings of the App.\n IMPORTANT: Delete the message above to preserve your password!</string>
+    <string name="oat_sms_message_error_feature_not_found">OAT: \n The requested feature %s does not exist. Ensure that you typed the name correctly.\n IMPORTANT: Delete the message above to preserve your password!</string>
+    <string name="oat_sms_message_send_gps_position">OAT: \n The device is currently located at the following coordinates: %f,%f\n IMPORTANT: Delete the message above to preserve your password!</string>
+    <string name="oat_sms_message_error_invalid_password">OAT: \n The password \"%s\" is invalid. Ensure that you typed the name correctly.</string>
```
Comment (tacticalDevC) [inline app/src/main/res/values/oat_sms_texts.xml]: Ouch. We shouldn't send the user his password even if it is wrong. And a password isn't a name. Please fix the typo(s)

Comment (FiryDragon1) [reply app/src/main/res/values/oat_sms_texts.xml]: Correct, fixed

File: app/src/main/java/at/tacticaldevc/oat/utils/SMSCom.java (hunk 2 above)
Comment (FiryDragon1) [reply app/src/main/java/at/tacticaldevc/oat/utils/SMSCom.java]: I created a new Exception:
OATApplicationException.forNoSMSSubscriptionFound()
This Error occurs if a phone has no Subscription to send SMS, as such it is not a Library Error.

File: app/src/main/java/at/tacticaldevc/oat/listeners/GPSListenerSMS.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    /**
+     * Called when the user enables the Provider
+     * Not used
+     *
+     * @param provider the Provider
+     */
+    @Override
+    public void onProviderDisabled(String provider) {
```
[the remark this answers is not in the document]
Comment (FiryDragon1) [reply app/src/main/java/at/tacticaldevc/oat/listeners/GPSListenerSMS.java]: The Provider is usually only disabled/enabled if the User does so in the Settings of the Device. I nevertheless changed the documentation.

File: app/src/main/java/at/tacticaldevc/oat/utils/Prefs.java (hunk 1 above)
Comment (FiryDragon1) [reply app/src/main/java/at/tacticaldevc/oat/utils/Prefs.java]: implemented change

**Pull request merged**